Add documentation for azure incremental sync (#2146)
port-labs/port-docs: docs/build-your-software-catalog/sync-data-to-catalog/cloud-providers/azure/incremental-sync.md

---
sidebar_position: 3
---

import Tabs from "@theme/Tabs"
import TabItem from "@theme/TabItem"
import CredentialsGuide from "/docs/build-your-software-catalog/custom-integration/api/\_template_docs/\_find_credentials.mdx";

# Incremental Sync

The Azure incremental sync script provides a way to periodically sync Azure resources, subscriptions, and resource groups into Port. This script is particularly useful when you want more control over the sync process or need to customize the data being synced.

## Overview

This script:
- Is written in Python using the Azure SDK for Python
- Runs as a GitHub workflow at configurable periodic intervals
- Provides two sync modes:
  - **Incremental**: Syncs recent changes (default: every 15 minutes)
  - **Full**: Complete sync of all resources (recommended for initial setup)
- Prioritizes Azure resources with detailed tracking
- Supports flexible data mapping through Port webhooks
- Can be deployed via:
  - GitHub Actions for automated periodic sync
  - Local installation for development and testing

> :::tip Repository

Reviewer: ```suggestion
:::tip Source code
```

The source code is available in the [port-labs/incremental-sync](https://github.com/port-labs/incremental-sync) repository.
:::

## Prerequisites

### Azure Setup

This integration requires the standard Azure [app registration](https://learn.microsoft.com/en-us/entra/identity-platform/quickstart-register-app?tabs=certificate%2Cexpose-a-web-api) setup described in the [installation guide](/build-your-software-catalog/sync-data-to-catalog/cloud-providers/azure/installation.md), plus some additional permissions.

Keep the following credentials handy after setup:
- `AZURE_CLIENT_ID`: The client ID of the Azure service principal
- `AZURE_CLIENT_SECRET`: The client secret of the Azure service principal
- `AZURE_TENANT_ID`: The tenant ID of the Azure service principal

#### Additional Permissions Required

In addition to the standard Azure App Registration setup, this integration requires:

> 1. **Azure Resource Graph Access**:
>    - In your app registration's "API permissions":
>    - Click "Add a permission"
>    - Click on "APIs my organization uses" tab
>    - Search for "Azure Resource Graph"
>    - Select "Read" permission
>    - Click "Add permissions"
>    - Click "Grant admin consent"

Reviewer: ```suggestion
1. **Azure Resource Graph Access**:
   In your app registration's "API permissions":
   - Click "Add a permission"
   - Click on "APIs my organization uses" tab
   - Search for "Azure Resource Graph"
   - Select "Read" permission
   - Click "Add permissions"
   - Click "Grant admin consent"
```

2. **Verification Steps**:
   Before proceeding, verify your setup can:
   - List your subscriptions
   - Query Azure Resource Graph
   - Read resource information

:::tip Troubleshooting Permissions
Common issues and solutions:
1. **Can't list subscriptions**: Verify Azure Service Management permission is granted
2. **Can't query resources**: Check Azure Resource Graph permission
3. **Access denied errors**: Ensure role assignment is correct at subscription level
4. **Intermittent failures**: Wait a few minutes after permission changes
:::

### Port Setup

The basic Port setup follows the [standard installation guide](/build-your-software-catalog/sync-data-to-catalog/cloud-providers/azure/installation.md#port-setup). However, this integration uses a different webhook configuration for incremental syncing:

#### Port Credentials

<CredentialsGuide />

#### Blueprint Configuration

:::tip Customizable Blueprints

> While we provide these configurations as a starting point, you can customize them based on your specific requirements and the Azure resources you want to track.

Reviewer: ```suggestion
While these configurations are provided as a starting point, you can customize them based on your specific requirements and the Azure resources you want to track.
```

      "target": "azureSubscription",
      "required": false,
      "many": false
    }
  }
}
```
</details>

<details>
<summary>azureCloudResources blueprint</summary>

```json
{
  "identifier": "azureCloudResources",
  "description": "This blueprint represents an AzureCloud Resource in our software catalog",
  "title": "Azure Cloud Resources",
  "icon": "Git",
  "schema": {
    "properties": {
      "tags": {
        "title": "Tags",
        "type": "object"
      },
      "type": {
        "title": "Type",
        "type": "string"
      },
      "location": {
        "title": "Location",
        "type": "string"
      }
    },
    "required": []
  },
  "mirrorProperties": {},
  "calculationProperties": {},
  "aggregationProperties": {},
  "relations": {
    "resourceGroup": {
      "title": "Resource Group",
      "target": "azureResourceGroup",
      "required": false,
      "many": false
    }
  }
}
```
</details>

#### Webhook Configuration

![Azure Basic Blueprints](../../../../../static/img/build-your-software-catalog/sync-data-to-catalog/cloud-providers/azure/create-webhook.png)

To set up the webhook in Port:

1. Navigate to the "Builder" page

> 2. Click on "Data Sources" in the left sidebar

Reviewer: when directing readers to pages in the app, use a link. In this case you can link directly to the data sources page, so the first 2 steps can be made into one.

:::

</TabItem>
</Tabs>

</TabItem>

<TabItem value="local" label="Local Installation">

To run the integration locally, follow these steps:

1. Clone the repository:
```bash
git clone https://github.com/port-labs/incremental-sync.git
cd integrations/azure_incremental
```

2. Install dependencies using Poetry:
```bash
pip install poetry
make install
```

3. Set the required environment variables:
```bash
# Required variables
export AZURE_CLIENT_ID="your-azure-client-id"
export AZURE_CLIENT_SECRET="your-azure-client-secret"
export AZURE_TENANT_ID="your-azure-tenant-id"
export PORT_WEBHOOK_INGEST_URL="your-port-webhook-url"

# Optional variables
export SUBSCRIPTION_BATCH_SIZE=1000  # Default: 1000 (max)
export CHANGE_WINDOW_MINUTES=15      # Default: 15 minutes
export RESOURCE_TYPES='["microsoft.keyvault/vaults","Microsoft.Network/virtualNetworks"]'  # Default: All
```

4. Run the integration:
```bash
# For incremental sync
make run

# For full sync
export SYNC_MODE=full
make run
```

:::tip
For development and testing purposes, you can use a smaller `CHANGE_WINDOW_MINUTES` value to sync changes more frequently.
:::

</TabItem>
</Tabs>

> ## How It Works

Reviewer: only the first word in headers should be capitalized, in this case it should be "How it works". please fix for all headers.
